### ISSUE TITLE
SpaczzRuler configuration 

### ISSUE THREAD

**Issue by lfoppiano** [gandersen101/spaczz#56]
Hi, 
   I could not find a way to set the various matching parameters using spacy 2, in the SpaczzRuler documentation https://github.com/gandersen101/spaczz/blob/master/examples/fuzzy_matching_tweaks.md there is only the spacy 3 syntax. 

I also found this https://github.com/gandersen101/spaczz/issues/18 and I tried to applied using the following: 

``ruler = SpaczzRuler(self.nlp, spaczz_fuzzy_defaults={'min_r2': 98, 'min_r1': 95, 'flex': 2})``

but changing parameters didn't seem to change anything. 

is that correct or I'm missing something? 

Thank you for the cool library!

**Comment by gandersen101** [comment]
Hi @Ifoppiano. Thanks for the feedback! Spaczz's internal syntax is mostly the same regardless of spaCy v2/3. The main difference is how spaCy v3 deals with adding pipeline components, so it looks like you are changing the fuzzy defaults correctly.

There's a decent amount of detail below but TL;DR: Make your `min_r` values more permissive and you'll see more meaningful results.

I believe you aren't seeing any changes in your results because your `min_r` values are too strict. Fuzzy matching ratios are affected by differences proportionally to the length of the compared strings, so unless you are dealing with very long strings, each difference is probably worth more than 2-5 ratio points.

See how a one character difference between these two 10-character strings makes a 10-point ratio difference using the default fuzzy matcher.


```python
from rapidfuzz import fuzz

int(fuzz.ratio("abcdefghij", "abcdefghii"))
```




    90



So using your `min_r` defaults in a SpaczzRuler with spaCy v2, we won't get a match.


```python
import spacy
from spaczz.pipeline import SpaczzRuler

nlp = spacy.blank("en")
ruler = SpaczzRuler(nlp, spaczz_fuzzy_defaults={"min_r2": 98, "min_r1": 95})
patterns = [
    {"label": "TEST", "pattern": "abcdefghij", "type": "fuzzy"}
]
ruler.add_patterns(patterns)

doc = nlp("abcdefghii")
doc = ruler(doc)
print(doc.ents)
```

    ()


But we can see that the defaults are in the ruler.


```python
ruler.defaults
```




    {'fuzzy_defaults': {'min_r2': 98, 'min_r1': 95}}



So let's make these values more permissive. We'll set `min_r2` to `90` and get rid of `min_r1` for this example.


```python
nlp = spacy.blank("en")
ruler = SpaczzRuler(nlp, spaczz_fuzzy_defaults={"min_r2": 90})
patterns = [
    {"label": "TEST", "pattern": "abcdefghij", "type": "fuzzy"}
]
ruler.add_patterns(patterns)

doc = nlp("abcdefghii")
doc = ruler(doc)
print(doc.ents)
```

    (abcdefghii,)


Here we get a match and verify that the defaults in the ruler reflect the changes we made.


```python
ruler.defaults
```




    {'fuzzy_defaults': {'min_r2': 90}}



A note on the `flex` parameter. This allows patterns to be compared against longer/shorter spans in the text (in terms of spaCy tokenization) than the pattern is itself.

For example see how the fuzzy ratio between these two strings is also 90.


```python
int(fuzz.ratio("abcdefghij", "abcd efghj"))
```




    90



However spaCy will tokenize the first string as one token and the second string as two. So the default `flex` value of `len(pattern) // 1`, or `0` in the below case, means the below pattern will only be compared against single tokens in the text resulting in no matches. This is what `min_r1` checks against in matches when `flex > 0`. If `flex == 0`, `min_r1` becomes unnecessary.


```python
nlp = spacy.blank("en")
ruler = SpaczzRuler(nlp, spaczz_fuzzy_defaults={"flex": "default"})
patterns = [
    {"label": "TEST", "pattern": "abcdefghij", "type": "fuzzy"}
]
ruler.add_patterns(patterns)
doc = nlp("abcd efghj")
doc = ruler(doc)
doc.ents
```




    ()



However, if we set `flex` to `"max"`, which is `len(pattern)`, or `1` in this case, it will search the surrounding 1 tokens of an initial match (what `min_r1` checks against) and see if a better match can be found. It'll then check the best match against `min_r2`. We see that setting `flex` to `"max"` instead allows us to identify the match. 


```python
nlp = spacy.blank("en")
ruler = SpaczzRuler(nlp, spaczz_fuzzy_defaults={"flex": "max"})
patterns = [
    {"label": "TEST", "pattern": "abcdefghij", "type": "fuzzy"}
]
ruler.add_patterns(patterns)
doc = nlp("abcd efghj")
doc = ruler(doc)
doc.ents
```




    (abcd efghj,)



I probably need to think about simplifying some of these API options and I have gone back and forth on what I want the `"min"`, `"default"` and `"max"` `flex` values to be but hopefully that gives you at least some clarity.

Let me know if you're still running into problems.
